File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.js
/**
 * // Definition for a _Node.
 * function _Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {_Node} head
 * @return {_Node}
 */
var copyRandomList = function(head) {
    const map = new Map();
    let i = 0;
    let temp = head;
    const res = new _Node(null, null, null);
    let tempres = res;
    while (temp) {
        const node = new _Node(temp.val, null, null);
        tempres.next = node;
        map.set(temp, node);
        tempres = tempres.next;
        temp = temp.next;
    }
    tempres = res.next;
    while (head) {
        const random = head.random;
        if (random !== null) {
            tempres.random = map.get(random);
        }
        head = head.next;
        tempres = tempres.next;
    }
    return res.next;
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        hmap = {}
        temp = head
        res = Node(0)
        tres = res
        while temp:
            node = Node(temp.val)
            hmap[temp] = node
            tres.next = node
            tres = tres.next
            temp = temp.next
        tres = res.next
        while head:
            if head.random:
                tres.random = hmap[head.random]
            head = head.next
            tres = tres.next
        return res.next

